src/music/midifile.cpp

```diff
@@ -450,8 +794,304 @@ void MidiFile::MoveFrom(MidiFile &other)
 	std::swap(this->tempos, other.tempos);
 	this->tickdiv = other.tickdiv;
 
+	_midifile_instance = this;
+
 	other.blocks.clear();
 	other.tempos.clear();
 	other.tickdiv = 0;
 }
 
+static void WriteVariableLen(FILE *f, uint32 value)
+{
+	if (value < 0x7F) {
+		byte tb = value;
+		fwrite(&tb, 1, 1, f);
+	} else if (value < 0x3FFF) {
+		byte tb[2];
+		tb[1] =  value & 0x7F;         value >>= 7;
+		tb[0] = (value & 0x7F) | 0x80; value >>= 7;
+		fwrite(tb, 1, sizeof(tb), f);
+	} else if (value < 0x1FFFFF) {
+		byte tb[3];
+		tb[2] =  value & 0x7F;         value >>= 7;
+		tb[1] = (value & 0x7F) | 0x80; value >>= 7;
+		tb[0] = (value & 0x7F) | 0x80; value >>= 7;
+		fwrite(tb, 1, sizeof(tb), f);
+	} else if (value < 0x0FFFFFFF) {
+		byte tb[4];
+		tb[3] =  value & 0x7F;         value >>= 7;
+		tb[2] = (value & 0x7F) | 0x80; value >>= 7;
+		tb[1] = (value & 0x7F) | 0x80; value >>= 7;
+		tb[0] = (value & 0x7F) | 0x80; value >>= 7;
+		fwrite(tb, 1, sizeof(tb), f);
+	}
+}
+
+/**
+ * Write a Standard MIDI File containing the decoded music.
+ * @param filename Name of file to write to
+ * @return True if the file was written to completion
+ */
+bool MidiFile::WriteSMF(const char *filename)
+{
+	FILE *f = FioFOpenFile(filename, "wb", Subdirectory::NO_DIRECTORY);
+	if (!f) {
+		return false;
+	}
+
+	uint16 u16;
+	byte bb;
+
+	/* SMF header */
+	const byte filemagic[] = { 'M', 'T', 'h', 'd', 0x00, 0x00, 0x00, 0x06 };
+	fwrite(filemagic, sizeof(filemagic), 1, f);
+	fwrite(&(u16 = 0), sizeof(u16), 1, f);
+	fwrite(&(u16 = TO_BE16(1)), sizeof(u16), 1, f);
+	fwrite(&(u16 = TO_BE16(this->tickdiv)), sizeof(u16), 1, f);
```

Comment:
neat trick, but looks terrifying - would be better split out, IMO. Alternatively - `(f)putc` ?

Same for several of the fwrites below